helpdesk/email.py

```diff
@@ -225,6 +226,101 @@ def imap_sync(q, logger, server):
     server.logout()
 
 
+def imap_oauth_sync(q, logger, server):
+    """
+    IMAP eMail server with OAUTH authentication.
+    Only tested against O365 implementation
+
+    Uses HELPDESK OAUTH Dict in Settings.
+
+    """
+
+    try:
+        logger.debug("Start Mailbox polling via IMAP OAUTH")
+
+        client = oauth2lib.BackendApplicationClient(
+            client_id=settings.HELPDESK_OAUTH["client_id"],
+            scope=settings.HELPDESK_OAUTH["scope"],
+        )
+
+        oauth = requests_oauthlib.OAuth2Session(client=client)
+        token = oauth.fetch_token(
+            token_url=settings.HELPDESK_OAUTH["token_url"],
+            client_id=settings.HELPDESK_OAUTH["client_id"],
+            client_secret=settings.HELPDESK_OAUTH["secret"],
+            include_client_id=True,
+        )
+
+        # TODO: Somehow link this to the debug level set within Django settings logging
+        server.debug = 4
+
+        # TODO: Perhaps store the authentication string template externally? Settings? Queue Table?
+        server.authenticate(
+            "XOAUTH2",
+            lambda x: f"user={q.email_box_user}\x01auth=Bearer {token['access_token']}\x01\x01".encode(),
+        )
+
+        # Select the Inbound Mailbox folder
+        server.select(q.email_box_imap_folder)
+
+    except imaplib.IMAP4.abort:
+        logger.error("IMAP authentication failed.")
+        server.logout()
+        sys.exit()
+
+    except ssl.SSLError:
+        logger.error(
+            "IMAP login failed due to SSL error. This is often due to a timeout. "
+            "Please check your connection and try again."
+        )
+        server.logout()
+        sys.exit()
```

Comment:
Change these excpetions to this format so that we get the exceptions reason for the problem rather than a generic and possibly misleading message. Adding **exc_info-True**  will also show the stack trace to aid in debugging:
```
    except imaplib.IMAP4.abort as e1:
        logger.error(f"IMAP authentication failed in OAUTH: {e1}", exc_info=True)
        server.logout()
        sys.exit()
    except ssl.SSLError as e2:
        logger.error(
            f"IMAP login failed due to SSL error. (This is often due to a timeout): {e2}", exc_info=True
        )
        server.logout()
        sys.exit()
```
I am not sure why they used sys.exit() in this module since it crashes the Django app and there is a more gracefull way to handle it so that it might recover but will not crash the entire Django app that Hlpdesk might be integrated into but this can be addressed in a separate PR.